Error messages with wrong alignment due to tab characters

When using tabs for indentation, error messages by the OCaml compiler appear with the wrong alignment:

```
File "common.ml", line 22, characters 36-55:
22 |            Some (Caml.really_input_string ch (Int32.to_int size))
                                         ^^^^^^^^^^^^^^^^^^^
Error: This expression has type int option
       but an expression was expected of type int
make: *** [Makefile:2: all] Error 1
```

This ends up becoming annoying if you always use tab characters, as it forces you to go to your text editor and manually go to the columns reported by the compiler to figure out which expression the error applies to.

--- a/.gitattributes
+++ b/.gitattributes
@@ -114,6 +114,7 @@ testsuite/tests/generated-parse-errors/errors.*         typo.very-long-line
 testsuite/tools/*.S                                     typo.missing-header
 testsuite/tools/*.asm                                   typo.missing-header
 testsuite/typing                                        typo.missing-header
+testsuite/tests/messages/highlight_tabs.ml              typo.tab
 
 # prune testsuite reference files
 testsuite/tests/**/*.reference               typo.prune

--- a/Changes
+++ b/Changes
@@ -85,6 +85,11 @@ Working version
   Makefile.config.
   (Damien Doligez, review by Mark Shinwell and Gabriel Scherer)
 
+- #9116, #9118, #10582: Fix single-line source highlighting in the
+  presence of tabs
+  (Armaël Guéneau, review by Gabriel Scherer,
+   split off from #9118 by Kate Deplaix, report by Ricardo M. Correia)
+
 ### Internal/compiler-libs changes:
 
 - #1599: add unset directive to ocamltest to clear environment variables before

--- a/parsing/location.ml
+++ b/parsing/location.ml
@@ -464,14 +464,20 @@ let highlight_quote ppf
         (* Single-line error *)
         Format.fprintf ppf "%s | %s@," line_nb line;
         Format.fprintf ppf "%*s   " (String.length line_nb) "";
-        for pos = line_start_cnum to rightmost.pos_cnum - 1 do
+        String.iteri (fun i c ->
+          let pos = line_start_cnum + i in
           if ISet.is_start iset ~pos <> None then
             Format.fprintf ppf "@{<%s>" highlight_tag;
           if ISet.mem iset ~pos then Format.pp_print_char ppf '^'
-          else Format.pp_print_char ppf ' ';
+          else if pos < rightmost.pos_cnum then begin
+            (* For alignment purposes, align using a tab for each tab in the
+               source code *)
+            if c = '\t' then Format.pp_print_char ppf '\t'
+            else Format.pp_print_char ppf ' '
+          end;
           if ISet.is_end iset ~pos <> None then
             Format.fprintf ppf "@}"
-        done;
+        ) line;
         Format.fprintf ppf "@}@,"
     | _ ->
         (* Multi-line error *)

--- a/testsuite/tests/messages/highlight_tabs.ml
+++ b/testsuite/tests/messages/highlight_tabs.ml
@@ -1,0 +1,13 @@
+(* TEST
+  * expect
+*)
+
+		let x = abc
+;;
+[%%expect{|
+Line 1, characters 10-13:
+1 | 		let x = abc
+              ^^^
+Error: Unbound value abc
+Hint: Did you mean abs?
+|}];;

--- a/testsuite/tests/messages/highlight_tabs.ml
+++ b/testsuite/tests/messages/highlight_tabs.ml
@@ -7,7 +7,7 @@
 [%%expect{|
 Line 1, characters 10-13:
 1 | 		let x = abc
-              ^^^
+    		        ^^^
 Error: Unbound value abc
 Hint: Did you mean abs?
 |}];;

